File: Canvas/TechDebt/Web Browser/WebBrowserViewController.m
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
    
    

#import <CanvasKit1/CanvasKit1.h>
#import "UIViewController+AnalyticsTracking.h"
#import <CanvasKit1/CKActionSheetWithBlocks.h>
#import "WebBrowserViewController.h"
#import "iCanvasConstants.h"
#import "Analytics.h"
#import "UIAlertController+TechDebt.h"

@import CanvasCore;
@import CanvasKit;

@interface WebBrowserViewController () <WKNavigationDelegate, UIDocumentInteractionControllerDelegate, UITextFieldDelegate, NSURLConnectionDataDelegate>
{
    __weak IBOutlet UINavigationBar *topToolbar;
    __weak IBOutlet UIToolbar *bottomToolbar;
    __weak IBOutlet UIBarButtonItem *backButton;
    __weak IBOutlet UIBarButtonItem *forwardButton;
    __weak IBOutlet UIBarButtonItem *refreshButton;
    __weak IBOutlet UIBarButtonItem *actionButton;
    __strong IBOutlet UIBarButtonItem *stopButton;
    NSUInteger networkOps;
    NSArray *reloadBarItems;
    NSArray *stopBarItems;
    CKActionSheetWithBlocks *optionsActionSheet;
    CKActionSheetWithBlocks *openInErrorSheet;
    UIDocumentInteractionController *dic;
}

@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (nonatomic, weak) IBOutlet UIView *webViewContainer;
@property (nonatomic, strong) CanvasWebView *webView;
@property (nonatomic, strong) NSString *mime;
@property (nonatomic, strong) NSURL *fileURLPath;
@property (weak, nonatomic) IBOutlet UITextField *titleField;
@property (nonatomic, strong) NSString *fullURL;
@property (nonatomic, strong) NSString *hostURL;
@property (nonatomic, assign) BOOL isTitleAbreviated;
@property (nonatomic, assign) BOOL isAnimatingTitle;

@end


@implementation WebBrowserViewController

- (instancetype)initWithURL:(NSURL *)url
{
    UINavigationController *nav = (UINavigationController *)[[UIStoryboard storyboardWithName:@"Storyboard-WebBrowser" bundle:[NSBundle bundleForClass:[self class]]] instantiateInitialViewController];
    self = nav.viewControllers[0];
    [self setUrl:url];
    return self;
}

- (instancetype)initWithURL:(NSURL *)url delegate:(id<WebBrowserViewControllerDelegate>)delegate
{
    self = [[WebBrowserViewController alloc] initWithURL:url];
    [self setDelegate:delegate];
    return self;
}

- (void)dealloc {
    _webView.navigationDelegate = nil;
    dic.delegate = nil;
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    if (self.fileURLPath) {
        NSError *error;
        [[NSFileManager defaultManager] removeItemAtURL:self.fileURLPath error:&error];
        if (error) {
            NSLog(@"Could not remove the file at that location: %@", self.fileURLPath);
        }
    }
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self configureWebView];

    if ([self presentingViewController] || self.browserWillDismissBlock) {
        self.doneButton.title = NSLocalizedString(@"Close", @"Close the window");
    } else {
        self.navigationItem.leftBarButtonItems = @[];
    }

    topToolbar.topItem.title = @"";
    
    // Set up our two states of buttons for the bottom toolbar.
    // The first one is just a copy of what is laid out in the xib
    // The second one swaps out the Refresh button for a Stop button
    reloadBarItems = bottomToolbar.items;
    NSMutableArray *tempButtons = [NSMutableArray new];
    for (UIBarButtonItem *item in reloadBarItems) {
        if (item == refreshButton) {
            if (! stopButton) {
                stopButton = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Stop",@"Stop loading button for webview") style:UIBarButtonItemStylePlain target:self action:@selector(stopButtonTapped:)];
            }
            [tempButtons addObject:stopButton];
        }
        else {
            [tempButtons addObject:item];
        }
    }
    stopBarItems = tempButtons;
    
    // Set up the error action sheet
    openInErrorSheet = [[CKActionSheetWithBlocks alloc] initWithTitle:NSLocalizedString(@"No installed apps support opening this file", @"Error message")];
    [openInErrorSheet addButtonWithTitle:NSLocalizedString(@"OK", nil)];
    
    if (self.request) {
        [self.request loadRequestInWebView:self.webView];
    }
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    [self.view setBackgroundColor:[UIColor prettyBlack]];
    [self.titleField setTextColor:[UIColor whiteColor]];
    [self.titleField setBackgroundColor:[UIColor clearColor]];
    [self setIsTitleAbreviated:YES];
    [self.titleField setBorderStyle:UITextBorderStyleNone];
    
    UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(webViewTapped:)];
    [self.webView addGestureRecognizer:tapRecognizer];
    
    [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(onKeyboardHide:) name:UIKeyboardWillHideNotification object:nil];
    
    [self.titleField setTintColor:Brand.current.tintColor];
//    [self.webView setScalesPageToFit:YES];
}

- (void)onKeyboardHide:(NSNotification *)notification
{
    if ( ! self.isTitleAbreviated) {
        [self toggleTitleDisplayState];
    }
}

- (void)webViewTapped:(id)sender
{
    if ( ! self.isTitleAbreviated) {
        [self toggleTitleDisplayState];
        [self.titleField resignFirstResponder];
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [self.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:self.titleField.text]]];
    [self.titleField resignFirstResponder];
    return YES;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    NSLog(@"TRACKING: %@", kGAIScreenWebBrowser);
    [Analytics logScreenView:kGAIScreenWebBrowser];
}

- (void)configureWebView
{
    self.webView = [CanvasWebView new];
    self.webView.navigationDelegate = self;
    [self.webViewContainer addSubview:self.webView];
    [self.webView pinToAllSidesOfSuperview];
}

#pragma mark - Setters

- (void)setUrl:(NSURL *)url {
    _url = url;
    
    if (url) {
        self.request = [NSURLRequest requestWithURL:url];
    
        [self.request loadRequestInWebView:self.webView];
    }
}

- (void)setContentHTML:(NSString *)html baseURL:(NSURL *)baseURL {
    self.request = [StaticHTMLRequest requestWithHTML:html baseURL:baseURL];
}

#pragma mark -
#pragma mark IBAction

- (IBAction)doneButtonTapped:(id)sender
{
    [[self presentingViewController] dismissViewControllerAnimated:YES completion:^{
        // Blank the webview by loading empty HTML
        [_webView loadHTMLString:@"" baseURL:nil];
    }];
    if (self.browserWillDismissBlock) {
        self.browserWillDismissBlock();
    }
}

- (IBAction)backButtonTapped:(id)sender
{
    if ([_webView canGoBack]) {
        networkOps = 0;
        [_webView goBack];
    }
    else {
        NSLog(@"Can't go back in the history.");
    }
}

- (IBAction)forwardButtonTapped:(id)sender
{
    if ([_webView canGoForward]) {
        networkOps = 0;
        [_webView goForward];
    }
    else {
        NSLog(@"Can't go forward in the history.");
    }
}

#pragma mark TextField Delegate


- (void)textFieldDidBeginEditing:(UITextField *)textField
{
    [self toggleTitleDisplayState];
}

- (IBAction)refreshButtonTapped:(id)sender
{
    networkOps = 0;
    if (self.webView.URL.absoluteString.length && ![self.webView.URL.absoluteString isEqualToString:@"about:blank"]) {
        [_webView reload];
    } else {
        [self.request loadRequestInWebView:self.webView];
    }
}

- (IBAction)stopButtonTapped:(id)sender
{
    networkOps = 0;
    [_webView stopLoading];
}

- (IBAction)actionButtonTapped:(id)sender
{
    if ([optionsActionSheet isVisible]) {
        [optionsActionSheet dismissWithClickedButtonIndex:[optionsActionSheet cancelButtonIndex] animated:NO];
    }
    
    if ([openInErrorSheet isVisible]) {
        [openInErrorSheet dismissWithClickedButtonIndex:[openInErrorSheet cancelButtonIndex] animated:NO];
    }

    [self.webView evaluateJavaScript:@"document.title" completionHandler:^(id result, NSError *error) {
        BOOL addedAtLeastOneButton = NO;
        optionsActionSheet = [[CKActionSheetWithBlocks alloc] initWithTitle:result];

        // Present an action sheet.
        // 1. Open in Safari
        NSURL *theURL = [NSURL URLWithString:self.fullURL]; // Copy to a local variable to avoid a retain cycle
        if (self.request.canOpenInSafari) {
            [optionsActionSheet addButtonWithTitle:NSLocalizedString(@"Open in Safari", @"Open a document in the application Safari") handler:^{
                [[UIApplication sharedApplication] openURL:theURL options:@{} completionHandler:nil];
            }];
            addedAtLeastOneButton = YES;
        }

        // 2. Open in another application
        CKActionSheetWithBlocks *localOpenInErrorSheet = openInErrorSheet;

        if (self.fileURLPath) {
            dic = [UIDocumentInteractionController interactionControllerWithURL:self.fileURLPath];
            dic.delegate = self;
            __weak UIDocumentInteractionController *weakDic = dic;
            [optionsActionSheet addButtonWithTitle:NSLocalizedString(@"Open in...", @"Open file in another application") handler:^{
                // Dismiss this action sheet and generate an OpenIn DIC display.
                // If there are no apps to handle the URL, display an action sheet that says, "Nothing supports this document". This is required to be in the App Store.
                BOOL presentedOpenInMenu = [weakDic presentOpenInMenuFromBarButtonItem:sender animated:YES];

                if (presentedOpenInMenu == NO) {
                    [localOpenInErrorSheet showFromBarButtonItem:sender animated:YES];
                }
            }];
            addedAtLeastOneButton = YES;
        }

        // 3. Print
        // 4. Make a sandwich
        // 5. Sudo make a sandwich
        // 6. Cancel
        [optionsActionSheet addCancelButtonWithTitle:NSLocalizedString(@"Cancel", nil)];;

        if (addedAtLeastOneButton == NO) {
            optionsActionSheet.title = NSLocalizedString(@"There are no actions for this item", @"Error message");
        }

        [optionsActionSheet showFromBarButtonItem:sender animated:YES];
    }];
}

#pragma mark -
#pragma mark UIDocumentInteractionControllerDelegate

- (void) documentInteractionController: (UIDocumentInteractionController *) controller didEndSendingToApplication: (NSString *) application
{
}

#pragma mark -
#pragma mark UIWebViewDelegate

- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
{
    self.fullURL = webView.URL.description;
    self.hostURL = webView.URL.host.description;

    if (self.isTitleAbreviated) {
        [self.titleField setText:self.hostURL];
    } else {
        [self.titleField setText:self.fullURL];
        [self toggleTitleDisplayState];
    }

    decisionHandler(WKNavigationActionPolicyAllow);
}

- (void)downloadFileAtURL:(NSURL *)url
{
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:configuration];
    
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    
    NSURLSessionDownloadTask *downloadTask = [manager downloadTaskWithRequest:request progress:nil destination:^NSURL *(NSURL *targetPath, NSURLResponse *response) {
        NSURL *documentsDirectoryPath = [NSURL fileURLWithPath:[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject]];
        return [documentsDirectoryPath URLByAppendingPathComponent:[response suggestedFilename]];
    } completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error) {
        self.fileURLPath = filePath;
    }];
    [downloadTask resume];
}

- (void)webView:(WKWebView *)webView didFailNavigation:(null_unspecified WKNavigation *)navigation withError:(NSError *)error;
{
    // Display an error page on the screen?
    if (networkOps > 0) networkOps--;
    if (networkOps == 0) {
        [self.activityIndicator stopAnimating];
        bottomToolbar.items = reloadBarItems;
        actionButton.enabled = YES;
    }
    
    if (error && [error code] != kCFURLErrorCancelled) {
        
        if ([error code] == 204 && [error userInfo][NSURLErrorFailingURLStringErrorKey]) {
            // Handle Kaltura media
            // 204 is "Plug-in handled load", meaning it was handled outside the webview. Just let it be.
            return;
        } else if ([error code] == 555) {
            NSLog(@"error: %@", error);
            
            //    MBL-5318 - If Penn State return an actual HTTP error code instead of swallowing the error then we can reload the page with the referrer for them - nlambson
            return;
        } else {
            NSString *localizedMessage = NSLocalizedString(@"There was an error loading the document.", @"Web browser error handling message");
            
            NSString *htmlMessage = [NSString stringWithFormat:@"<html><body style=\"font-family:sans-serif;font-size:30px;text-align:center;color:#555;padding:5px;\">%@</body></html>", localizedMessage];
            
            [_webView loadHTMLString:htmlMessage baseURL:nil];
        }
    }
}

- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(null_unspecified WKNavigation *)navigation
{
    networkOps++;
    
    backButton.enabled = [_webView canGoBack];
    forwardButton.enabled = [_webView canGoForward];
    
    if (networkOps == 1) {
        [self.activityIndicator startAnimating];
        bottomToolbar.items = stopBarItems;
        actionButton.enabled = NO;
    }
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(null_unspecified WKNavigation *)navigation
{
    backButton.enabled = [_webView canGoBack];
    [self setFullURL:webView.URL.description];
    // Show the title of the document.
    if ([webView.URL isFileURL]) {
        topToolbar.topItem.title = [webView.URL lastPathComponent];
    }
    else {
        __weak typeof(self) weakself = self;
        [webView evaluateJavaScript:@"document.title" completionHandler:^(id _Nullable result, NSError *_Nullable error) {
            [weakself.titleField setText:result];
            [weakself setHostURL:result];
        }];
    }
    
    if (networkOps > 0) networkOps--;
    if (networkOps == 0) {
        [self.activityIndicator stopAnimating];
        bottomToolbar.items = reloadBarItems;
        actionButton.enabled = YES;
    }
    
    [UIView animateWithDuration:0.3 animations:^{
        if (self.titleField.text.length) {
            CGSize size = [self.titleField.text sizeWithAttributes:@{NSFontAttributeName: self.titleField.font}];
            [self.titleField setFrame:CGRectMake((self.view.frame.size.width - roundf(size.width)) * 0.5, self.titleField.frame.origin.y, roundf(size.width), self.titleField.frame.size.height)];
        }
    }];

    [webView evaluateJavaScript:@"document.body.innerHTML" completionHandler:^(id _Nullable html, NSError *_Nullable error) {
        if ([html isEqualToString:@"Could not find download URL"]) {
            [UIAlertController showAlertWithTitle:NSLocalizedString(@"Error", nil) message:NSLocalizedString(@"There was an error loading your content. If it is an audio or video upload it may still be processing.", nil) handler:^{
                [self dismissViewControllerAnimated:YES completion:nil];
            }];
        }
    }];

    [webView replaceHREFsWithAPISafeURLs];
    [self.delegate webBrowser:self didFinishLoadingWebView:webView];
}

- (void)toggleTitleDisplayState
{
    if (self.isAnimatingTitle) {
        return;
    }
    if (self.isTitleAbreviated) {
        [self setIsAnimatingTitle:YES];
        [self.titleField setText:self.fullURL];
        [self.titleField setBorderStyle:UITextBorderStyleRoundedRect];
        [UIView animateWithDuration:0.3 animations:^{
            [self.titleField setTextColor:[UIColor darkTextColor]];
            [self.titleField setFrame:CGRectMake(0, self.titleField.frame.origin.y, self.view.frame.size.width, self.titleField.frame.size.height)];
            [self.titleField setBackgroundColor:[UIColor whiteColor]];
        } completion:^(BOOL finished) {
            [self setIsAnimatingTitle:NO];
        }];
    } else {
        [self setIsAnimatingTitle:YES];
        [UIView animateWithDuration:0.3 animations:^{
            [self.titleField setText:self.hostURL];
            CGSize size = [self.hostURL sizeWithAttributes:@{NSFontAttributeName: self.titleField.font}];
            [self.titleField setFrame:CGRectMake((self.view.frame.size.width - roundf(size.width)) * 0.5, self.titleField.frame.origin.y, roundf(size.width), self.titleField.frame.size.height)];
            [self.titleField setBackgroundColor:[UIColor clearColor]];
            [self.titleField setTextColor:[UIColor whiteColor]];
        } completion:^(BOOL finished) {
            [self setIsAnimatingTitle:NO];
            [self.titleField setBorderStyle:UITextBorderStyleNone];
        }];
        [self.titleField resignFirstResponder];
    }
    
    [self setIsTitleAbreviated:! self.isTitleAbreviated];
}


@end


@implementation NSURLRequest (WebBrowser)

- (void)loadRequestInWebView:(WKWebView *)webView
{
    if (self.URL != nil && self.URL.isFileURL) {
        [webView loadFileURL:self.URL allowingReadAccessToURL:self.URL];
    } else {
        [webView loadRequest:self];
    }
}

- (BOOL)canOpenInSafari {
    return !self.URL.isFileURL;
}

@end

@interface StaticHTMLRequest ()
@property (nonatomic, nonnull) NSString *html;
@property (nonatomic, nonnull) NSURL *baseURL;
@end

@implementation StaticHTMLRequest

+ (instancetype)requestWithHTML:( NSString * _Nonnull )html baseURL:(NSURL * _Nonnull)baseURL {
    StaticHTMLRequest *req = [self new];
    req.html = html;
    req.baseURL = baseURL;
    return req;
}

- (void)loadRequestInWebView:(WKWebView *)webView
{
    [webView loadHTMLString:self.html baseURL:self.baseURL];
}

- (BOOL)canOpenInSafari {
    return NO;
}

@end

File: Parent/Parent/Browser/WebBrowserViewController.swift
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import UIKit
import CFNetwork
import CanvasCore

class WebBrowserViewController: UIViewController {

    // MARK: Important stuffs
    var request: URLRequest?
    var url: URL? {
        didSet {
            if let url = url {
                self.request = URLRequest.requestWithDefaultHTTPHeaders(url)
            }
        }
    }

    let useAPISafeLinks: Bool
    let isModal: Bool

    // MARK: Private stuffs
    // MARK: Outlets
    fileprivate var webView = UIWebView()
    fileprivate var titleField = UITextField()

    fileprivate var doneButton: UIBarButtonItem!
    fileprivate var reloadButton: UIBarButtonItem!
    fileprivate var activityItem: UIBarButtonItem!
    fileprivate var stopButton: UIBarButtonItem!
    fileprivate var backButton: UIBarButtonItem!
    fileprivate var forwardButton: UIBarButtonItem!
    fileprivate var actionButton: UIBarButtonItem!

    fileprivate var activityIndicator: UIActivityIndicatorView = UIActivityIndicatorView(activityIndicatorStyle: .gray)

    fileprivate var networkOps: UInt = 0
    fileprivate var fullURLString: String?
    fileprivate var hostURLString: String?
    fileprivate var fileURLString: String?
    fileprivate var isTitleAbreviated: Bool = true
    fileprivate var isAnimatingTitle: Bool = false

    init(useAPISafeLinks: Bool = true, isModal: Bool = true) {
        self.useAPISafeLinks = useAPISafeLinks
        self.isModal = isModal
        super.init(nibName: nil, bundle: nil)

        self.doneButton = UIBarButtonItem(title: NSLocalizedString("Close", comment: "Close Button Title"), style: .plain, target: self, action: #selector(WebBrowserViewController.doneButtonTapped(_:)))
        self.reloadButton = UIBarButtonItem(barButtonSystemItem: .refresh, target: self, action: #selector(WebBrowserViewController.reloadButtonTapped(_:)))
        self.activityItem = UIBarButtonItem(customView: activityIndicator)
        self.stopButton = UIBarButtonItem(barButtonSystemItem: .stop, target: self, action: #selector(WebBrowserViewController.stopButtonTapped(_:)))
        self.backButton = UIBarButtonItem(image: UIImage(named: "back"), style: .plain, target: self, action: #selector(WebBrowserViewController.backButtonTapped(_:)))
        self.forwardButton = UIBarButtonItem(image: UIImage(named: "forward"), style: .plain, target: self, action: #selector(WebBrowserViewController.forwardButtonTapped(_:)))
        self.actionButton = UIBarButtonItem(barButtonSystemItem: .action, target: self, action: #selector(WebBrowserViewController.actionButtonTapped(_:)))

        let fixedSpace = UIBarButtonItem(barButtonSystemItem: .fixedSpace, target: nil, action: nil)
        fixedSpace.width = 20
        self.toolbarItems = [backButton, fixedSpace, forwardButton, UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), actionButton]
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        if let fileURLString = fileURLString {
            do {
                try FileManager.default.removeItem(atPath: fileURLString)
            } catch {
                print(error)
            }
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        navigationController?.isToolbarHidden = false
        if isModal {
            navigationItem.leftBarButtonItem = doneButton
        }
        navigationItem.rightBarButtonItems = [reloadButton]
        navigationItem.title = ""
        navigationItem.titleView = titleField

        webView.delegate = self
        webView.translatesAutoresizingMaskIntoConstraints = false
        webView.scalesPageToFit = true
        view.addSubview(webView)

        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[webView]|", options: NSLayoutFormatOptions(rawValue: 0), metrics: nil, views: ["webView": webView]))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[webView]|", options: NSLayoutFormatOptions(rawValue: 0), metrics: nil, views: ["webView": webView]))

        if let request = self.request {
            webView.loadRequest(request)
        }
        edgesForExtendedLayout = UIRectEdge()

        view.backgroundColor = UIColor.prettyBlack()
        titleField.textColor = self.navigationController?.navigationBar.barStyle == .black ? UIColor.white : UIColor.darkText
        titleField.backgroundColor = UIColor.clear
        titleField.borderStyle = .none
        titleField.tintColor = Brand.current.tintColor
        titleField.returnKeyType = .go
        titleField.delegate = self

        var frame = titleField.frame
        frame.size.height = 30
        titleField.frame = frame

        NotificationCenter.default.addObserver(self, selector: #selector(WebBrowserViewController.keyboardWillHide(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)

        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(WebBrowserViewController.webViewTapped(_:)))
        webView.addGestureRecognizer(tapGesture)
    }

    // MARK: Handlers
    func doneButtonTapped(_ doneButton: UIBarButtonItem) {
        self.presentingViewController?.dismiss(animated: true) { [weak self] in
            self?.webView.loadHTMLString("", baseURL: nil)
        }
    }

    func backButtonTapped(_ backButton: UIBarButtonItem) {
        guard webView.canGoBack else { return }

        networkOps = 0
        webView.goBack()
    }

    func forwardButtonTapped(_ forwardButton: UIBarButtonItem) {
        guard webView.canGoForward else { return }

        networkOps = 0
        webView.goForward()
    }

    func reloadButtonTapped(_ refreshButton: UIBarButtonItem) {
        networkOps = 0
        if let request = webView.request, (request.url?.absoluteString.characters.count)! > 0 && request.url?.absoluteString != "about:blank" {
            webView.reload()
        } else {
            if let request = self.request {
                webView.loadRequest(request)
            }
        }
    }

    func actionButtonTapped(_ actionButton: UIBarButtonItem) {
        let title = webView.stringByEvaluatingJavaScript(from: "document.title")
        let actionSheet = UIAlertController(title: title, message: nil, preferredStyle: .actionSheet)
        if let fullURLString = fullURLString, let url = URL(string: fullURLString), request?.url?.isFileURL == false {
            actionSheet.addAction(UIAlertAction(title: NSLocalizedString("Open in Safari", comment: "Open a url in the application Safari"), style: .default) { action in
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            })
        }

        if let fileURLString = fileURLString, let fileURL = URL(string: fileURLString) {
            let dic = UIDocumentInteractionController(url: fileURL)
            actionSheet.addAction(UIAlertAction(title: NSLocalizedString("Open in...", comment: "Open file in another application"), style: .default) { [weak dic, weak self] action in
                let presentedOpenInMenu = dic?.presentOpenInMenu(from: actionButton, animated: true)
                if presentedOpenInMenu == false {
                    let errorSheet = UIAlertController(title: NSLocalizedString("No installed apps support opening this file", comment: "Error message"), message: nil, preferredStyle: .actionSheet)
                    errorSheet.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "OK Button Title"), style: .default) { _ in })
                    self?.present(errorSheet, animated: true, completion: nil)
                }
            })
        }

        if actionSheet.actions.count == 0  {
            actionSheet.title = NSLocalizedString("There are no actions for this item", comment: "Error message")
        }

        actionSheet.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: "Cancel button title"), style: .cancel) { _ in })

        actionSheet.popoverPresentationController?.sourceView = self.view
        actionSheet.popoverPresentationController?.barButtonItem = actionButton
        present(actionSheet, animated: true, completion: nil)
    }

    func stopButtonTapped(_ stopButton: UIBarButtonItem) {
        networkOps = 0
        webView.stopLoading()
    }

    func webViewTapped(_ tapGesture: UITapGestureRecognizer) {
        if !isTitleAbreviated {
            toggleTitleDisplayState()
            titleField.resignFirstResponder()
        }
    }
}

// MARK: - UIWebViewDelegate
extension WebBrowserViewController: UIWebViewDelegate {
    func webView(_ webView: UIWebView, shouldStartLoadWith request: URLRequest, navigationType: UIWebViewNavigationType) -> Bool {
        fullURLString = request.url?.absoluteString
        hostURLString = request.url?.host

        if isTitleAbreviated {
            titleField.text = hostURLString
        } else {
            titleField.text = fullURLString
            toggleTitleDisplayState()
        }

        // TODO: URLConnection???

        return true
    }

    func webViewDidStartLoad(_ webView: UIWebView) {
        networkOps += 1

        backButton.isEnabled = webView.canGoBack
        forwardButton.isEnabled = webView.canGoForward

        if networkOps == 1 {
            activityIndicator.startAnimating()
            navigationItem.rightBarButtonItems = [stopButton, activityItem]
            actionButton.isEnabled = false
        }
    }

    func webViewDidFinishLoad(_ webView: UIWebView) {
        backButton.isEnabled = webView.canGoBack
        fullURLString = webView.request?.url?.absoluteString
        // Show the title of the document.
        if let url = webView.request?.url, url.isFileURL == true {
            navigationItem.title = webView.request?.url?.lastPathComponent
        } else {
            let title = webView.stringByEvaluatingJavaScript(from: "document.title")
            titleField.text = title
            hostURLString = title
        }

        if networkOps > 0 { networkOps -= 1 }
        if networkOps == 0 {
            activityIndicator.stopAnimating()
            navigationItem.rightBarButtonItems = [reloadButton]
            actionButton.isEnabled = true
        }

        UIView.animate(withDuration: 0.3, animations: {
            if let text = self.titleField.text, let font = self.titleField.font, text.characters.count > 0 {
                let size = (text as NSString).size(attributes: [NSFontAttributeName: font])
                let x: CGFloat = (self.view.frame.size.width - CGFloat(roundf(Float(size.width))) * 0.5)
                let width: CGFloat = CGFloat(roundf(Float(size.width)))
                self.titleField.frame = CGRect(x: x, y: self.titleField.frame.origin.y, width: width, height: self.titleField.frame.size.height)
            }
        }) 

        let html = webView.stringByEvaluatingJavaScript(from: "document.body.innerHTML")
        if html == "Could not find download URL" {
            let alert = UIAlertController(title: NSLocalizedString("Error", comment: ""), message: NSLocalizedString("There was an error loading your content. If it  is an audio or video upload it may still be processing.", comment: ""), preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .default, handler: nil))
            present(alert, animated: true, completion: nil)
        }

        if useAPISafeLinks {
            webView.replaceHREFsWithAPISafeURLs()
        }
    }

    func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
        let error = error as NSError
        if networkOps > 0 { networkOps -= 1 }
        if networkOps == 0 {
            activityIndicator.stopAnimating()
            navigationItem.rightBarButtonItems = [reloadButton]
            actionButton.isEnabled = true
        }

        if error.code != Int(CFNetworkErrors.cfurlErrorCancelled.rawValue) {
            if error.code == 204 && error.userInfo[NSURLErrorFailingURLErrorKey] != nil {
                // Handle Kaltura media
                // 204 is "Plug-in handled load", meaning it was handled outside the webview. Just let it be.
                return
            } else {
                webView.loadHTMLString("<html><body style=\"font-family:sans-serif;font-size:30px;text-align:center;color:#555;padding:5px;\">There was an error loading the document.</body></html>", baseURL: nil)
            }
        }
    }

    func downloadFile(atURL url: URL) {
        let config = URLSessionConfiguration.default
        let session = URLSession(configuration: config)
        let downloadTask = session.dataTask(with: url) { data, response, error in
            let filename = response?.suggestedFilename ?? "file"
            let url = URL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]).appendingPathComponent(filename)
            let _ = try? data?.write(to: url, options: .atomic)
            self.fileURLString = url.absoluteString
        }
        downloadTask.resume()
    }
}

// MARK: - UITextFieldDelegate
extension WebBrowserViewController: UITextFieldDelegate {
    func textFieldDidBeginEditing(_ textField: UITextField) {
        toggleTitleDisplayState()
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if let text = titleField.text, let url = URL(string: text) {
            self.url = url
            if let request = self.request {
                webView.loadRequest(request)
            }
        }
        titleField.resignFirstResponder()
        return true
    }

    func toggleTitleDisplayState() {
        guard !isAnimatingTitle else { return }
        defer { isTitleAbreviated = !isTitleAbreviated }

        isAnimatingTitle = true

        if isTitleAbreviated {
            titleField.text = fullURLString
            titleField.borderStyle = .roundedRect
            UIView.animate(withDuration: 0.3, animations: {
                self.titleField.frame = CGRect(x: 0, y: self.titleField.frame.origin.y, width: self.view.frame.size.width, height: self.titleField.frame.size.height)
                self.titleField.backgroundColor = UIColor.white
                self.titleField.textColor = UIColor.darkText
            }, completion: { _ in
                self.isAnimatingTitle = false
            })
        } else {
            titleField.text = hostURLString
            UIView.animate(withDuration: 0.3, animations: {
                if let hostURLString = self.hostURLString, let font = self.titleField.font {
                    let size = (hostURLString as NSString).size(attributes: [NSFontAttributeName: font])
                    let x: CGFloat = (self.view.frame.size.width - CGFloat(roundf(Float(size.width))) * 0.5)
                    let width: CGFloat = CGFloat(roundf(Float(size.width)))
                    self.titleField.frame = CGRect(x: x, y: self.titleField.frame.origin.y, width: width, height: self.titleField.frame.size.height)
                    self.titleField.backgroundColor = UIColor.clear
                    self.titleField.textColor = self.navigationController?.navigationBar.barStyle == .black ? UIColor.white : UIColor.darkText
                }
            }, completion: { _ in
                self.isAnimatingTitle = false
                self.titleField.borderStyle = .none
            })

            self.titleField.resignFirstResponder()
        }
    }
}

// MARK: - Keyboard Notifications
extension WebBrowserViewController {
    func keyboardWillHide(_ note: Notification) {
        if !isTitleAbreviated {
            toggleTitleDisplayState()
        }
    }
}
